includes/ical-parser/class-coblocks-ical.php

```diff
@@ -0,0 +1,2782 @@
+
+			// Process properties
+			if ( preg_match( '/([A-Z-]+)[;]([\w\W]*)/', $matches[0], $properties ) ) {
+				// Remove first match
+				array_shift( $properties );
+				// Fix to ignore everything in keyword after a ; (e.g. Language, TZID, etc.)
+				$matches[0] = $properties[0];
+				array_shift( $properties ); // Repeat removing first match
+
+				$formatted = array();
+				foreach ( $properties as $property ) {
+					// Match semicolon separator outside of quoted substrings
+					preg_match_all( '~[^' . PHP_EOL . '";]+(?:"[^"\\\]*(?:\\\.[^"\\\]*)*"[^' . PHP_EOL . '";]*)*~', $property, $attributes );
+					// Remove multi-dimensional array and use the first key
+					$attributes = ( sizeof( $attributes ) === 0 ) ? array( $property ) : reset( $attributes );
+
+					if ( is_array( $attributes ) ) {
+						foreach ( $attributes as $attribute ) {
+							// Match equals sign separator outside of quoted substrings
+							preg_match_all(
+								'~[^' . PHP_EOL . '"=]+(?:"[^"\\\]*(?:\\\.[^"\\\]*)*"[^' . PHP_EOL . '"=]*)*~',
+								$attribute,
+								$values
+							);
+							// Remove multi-dimensional array and use the first key
+							$value = ( sizeof( $values ) === 0 ) ? null : reset( $values );
+
+							if ( is_array( $value ) && isset( $value[1] ) ) {
+								// Remove double quotes from beginning and end only
+								$formatted[ $value[0] ] = trim( $value[1], '"' );
+							}
+						}
+					}
+				}
+
+				// Assign the keyword property information
+				$properties[0] = $formatted;
+
+				// Add match to beginning of array
+				array_unshift( $properties, $matches[1] );
+				$matches[1] = $properties;
+			}
+
+			return $matches;
+		} else {
+			return false; // Ignore this match
+		}
+	}
+
+	/**
+	 * Returns a `DateTime` object from an iCal date time format
+	 *
+	 * @param  string $ical_date
+	 * @return \DateTime
+	 * @throws \Exception
+	 */
+	public function ical_date_to_date_time( $ical_date ) {
+		/**
+		 * iCal times may be in 3 formats, (https://www.kanzaki.com/docs/ical/dateTime.html)
+		 *
+		 * UTC:      Has a trailing 'Z'
+		 * Floating: No time zone reference specified, no trailing 'Z', use local time
+		 * TZID:     Set time zone as specified
+		 *
+		 * Use DateTime class objects to get around limitations with `mktime` and `gmmktime`.
+		 * Must have a local time zone set to process floating times.
+		 */
+		$pattern  = '/^(?:TZID=)?([^:]*|".*")'; // [1]: Time zone
+		$pattern .= ':?';                       //      Time zone delimiter
+		$pattern .= '([0-9]{8})';               // [2]: YYYYMMDD
+		$pattern .= 'T?';                       //      Time delimiter
+		$pattern .= '(?(?<=T)([0-9]{6}))';      // [3]: HHMMSS (filled if delimiter present)
+		$pattern .= '(Z?)/';                    // [4]: UTC flag
+
+		preg_match( $pattern, $ical_date, $date );
+
+		if ( empty( $date ) ) {
+			throw new \Exception( 'Invalid iCal date format.' );
```

Comment:
In the context of WordPress we should be using [`WP_Error`](https://developer.wordpress.org/reference/classes/wp_error/) for error reporting.